File: python/lesson 4/quick_select.py
import random
from typing import List

# https://leetcode.com/problems/kth-largest-element-in-an-array
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        self.shuffle(nums)
        find = len(nums) - k
        lo = 0
        hi = len(nums) - 1
        while lo < hi:
            pivot = self.partition(nums, lo, hi)
            if pivot > find:
                hi = pivot - 1
            elif pivot < find:
                lo = pivot + 1
            else:
                break
        return nums[find]

    def partition(self, nums: List[int], lo: int, hi: int):
        pivot = nums[lo]
        i = lo
        j = hi + 1
        while True:
            while True:
                i += 1
                if not (nums[i] < pivot and i < hi): break
            while True:
                j -= 1
                if not (nums[j] > pivot and j > lo): break
            if i >= j: break
            self.swap(nums, i, j)
        self.swap(nums, lo, j)
        return j

    def swap(self, nums: List[int], a: int, b: int):
        nums[a], nums[b] = nums[b], nums[a]

    def shuffle(self, nums):
        for i in range(len(nums)):
            r = random.randrange(0, i + 1)
            self.swap(nums, i, r)

File: ruby/lesson 4/quick_select.rb
# https://leetcode.com/problems/kth-largest-element-in-an-array/

def find_kth_largest(nums, k)
    shuffle(nums)
    find = nums.length - k
    lo = 0
    hi = nums.length - 1
    while lo < hi
        pivot = partition(nums, lo, hi)
        if pivot > find
            hi = pivot - 1
        elsif pivot < find
            lo = pivot + 1
        else
            break
        end
    end
    nums[find]
end

def partition(nums, lo, hi)
    pivot = nums[lo]
    i = lo
    j = hi + 1
    loop do
        loop do
            i += 1
            break unless nums[i] < pivot && i < hi
        end
        loop do
            j -= 1
            break unless nums[j] > pivot && j > lo
        end
        break if i >= j
        swap(nums, i, j)
    end
    swap(nums, lo, j)
    j
end

def swap(nums, a, b)
    nums[a], nums[b] = nums[b], nums[a]
end

def shuffle(nums)
    (0...nums.length).each do |i|
        r = rand(0...i + 1)
        swap(nums, i, r)
    end
end
